src/mlpack/methods/reinforcement_learning/environment/maze.hpp

```diff
@@ -0,0 +1,406 @@
+
+  }
+
+  /**
+   * Generate Maze using Random Iterative DFS Algorithm with 
+   * 0.4 * rows * column steps
+   * @param rows number of rows in maze.
+   * @param columns number of columns in maze.
+   * 
+   */
+  void MazeGeneration(const size_t rows, const size_t columns)
+  {
+    maze = arma::mat(rows,columns);
+    arma::umat visited;
+    visited.zeros(rows, columns);
+
+    std::vector<arma::ivec> moves = {arma::ivec({0, -1}), arma::ivec({0, 1}), arma::ivec({-1, 0}), arma::ivec({1, 0}) };
+
+    const size_t maxNumberOfMazeSteps = 2 * std::sqrt(rows * columns);
+
+    std::stack<arma::ivec> cellsInPath;
+
+    arma::ivec startingPoint = {arma::randi(arma::distr_param(0, rows-1)), arma::randi(arma::distr_param(0, columns-1))};
+
+    cellsInPath.push(startingPoint);
+
+    size_t numberOfMazeSteps = 0;
+ 
+    // loop till stack is empty
+    while (!cellsInPath.empty())
+    {
+        // Pop a vertex from the stack
+        arma::ivec lastCell = cellsInPath.top();
+        cellsInPath.pop();
+
+        ++numberOfMazeSteps;
+
+        // Break for Max number of maze steps.
+        if(numberOfMazeSteps > maxNumberOfMazeSteps)
+          break;
+ 
+        // Mark the particular row and column as visited.
+        visited(lastCell(0),lastCell(1)) = 1;
+
+        // If visitable nodes are present from that node.
+        bool visitableNodes = false;
+
+        // Randomly choose a direction and check if the adjacent node in 
+        // that direction is visited or not.
+        for (size_t counter = 0; counter < 10 ; ++counter)
+        {    
+            size_t index = arma::randi(arma::distr_param(0, 3));
+            arma::ivec nextCell = arma::ivec({lastCell(0) + moves[index](0), {lastCell(1) + moves[index](1)}});
+            
+            // Check if particular row and column index are out of bounds of maze.
+            bool outOfBounds = nextCell(0)>= maze.n_rows || nextCell(1)>= maze.n_cols
+        || nextCell(0) < 0 || nextCell(1) < 0 ;
+            if (!outOfBounds && !visited(nextCell(0),nextCell(1))) 
+            {
+                cellsInPath.push(lastCell);
+                cellsInPath.push(nextCell);
+                break;
+            }
+        }
+    }
+
+    // Pop last ({row,column}) from the stack
+      arma::ivec goalCell = cellsInPath.top();
+      cellsInPath.pop();
+
+    // Set the ({row,column}) asa goal cell
+      maze(goalCell(0),goalCell(1)) = 1;
+
+    // Set all other cells in stack to 0 
+    while (!cellsInPath.empty())
+    {
+        arma::ivec lastCell = cellsInPath.top();
+        cellsInPath.pop();
+      
+        maze(lastCell(0),lastCell(1)) = 0;
+    }
+
+    // Randomly set other cells to -1(wall) or 0(path)
+    for (size_t row = 0; row < rows; ++row)
+    {
+      for (size_t col = 0; col < columns; ++col)
+      {
+          if(maze(row,col) !=0 && maze(row,col) !=1 )
+          {
+            maze(row,col) = arma::randi(arma::distr_param(-1, 0));
+          }
+      }
+    }
```

Comment:
Pretty sure we can use `stl` here or [imbue](https://arma.sourceforge.net/docs.html#imbue)

Reply:
```
std::generate(maze.begin(), maze.end(), 
  [](auto& elem) {
      if (elem != 0 && elem != 1)
           elem = arma::randi(arma::distr_param(-1, 0));
       });
```

OR

```
maze.transform([] (double cell) {
  if (cell != 0 && cell != 1)
     cell = arma::randi(arma::distr_param(-1, 0));
     });
```

Reply:
Sure that looks way better. I will make that change